apps/application-system/api/src/app/modules/application/application.controller.ts

```diff
@@ -404,86 +426,203 @@ export class ApplicationController {
     }
 
-    const [hasChanged, updatedApplication] = await this.changeState(
+    const {
+      hasChanged,
+      hasError,
+      error,
+      application: updatedApplication,
+    } = await this.changeState(
       mergedApplication,
       template,
       updateApplicationStateDto.event,
       authorization,
     )
 
-    // TODO: should not have to specificially check for updatedApplication
-    // because of return type on this.changeState
-    if (hasChanged === true && updatedApplication) {
+    if (hasError) {
+      throw new BadRequestException(error)
+    }
+
+    if (hasChanged) {
       return updatedApplication
     }
 
     return existingApplication
   }
 
-  async changeState(
+  async performActionOnApplication(
     application: BaseApplication,
     template: Unwrap<typeof getApplicationTemplateByTypeId>,
-    event: string,
     authorization: string,
-  ): Promise<[false] | [true, BaseApplication]> {
-    const helper = new ApplicationTemplateHelper(application, template)
+    action: ApplicationTemplateAPIAction,
+  ): Promise<TemplateAPIModuleActionResult> {
+    const {
+      apiModuleAction,
+      shouldPersistToExternalData,
+      externalDataId,
+      throwOnError,
+    } = action
+
+    const actionResult = await this.templateAPIService.performAction({
+      templateId: template.type,
+      type: apiModuleAction,
+      props: {
+        application,
+        authorization,
+      },
+    })
 
-    const [hasChanged, newState, newApplication] = helper.changeState(event)
+    let updatedApplication: BaseApplication = application
 
-    if (!hasChanged) {
-      return [false]
+    if (shouldPersistToExternalData) {
+      const newExternalDataEntry: ExternalData = {
+        [externalDataId || apiModuleAction]: {
+          status: actionResult.success ? 'success' : 'failure',
+          date: new Date(),
+          data: actionResult.success
+            ? (actionResult.response as ExternalData['data'])
+            : actionResult.error,
+        },
+      }
+
+      const {
+        updatedApplication: withExternalData,
+      } = await this.applicationService.updateExternalData(
+        updatedApplication.id,
+        updatedApplication.externalData,
+        newExternalDataEntry,
+      )
+
+      updatedApplication = {
+        ...updatedApplication,
+        externalData: {
+          ...updatedApplication.externalData,
+          ...withExternalData.externalData,
+        },
+      }
     }
 
-    const update = await this.applicationService.updateApplicationState(
-      application.id,
-      newState,
-      newApplication.answers,
-      newApplication.assignees,
-    )
+    if (!actionResult.success && throwOnError) {
+      return {
+        updatedApplication,
+        hasError: true,
+        error: actionResult.error,
+      }
+    }
 
-    const updatedApplication = update.updatedApplication as BaseApplication
-    const newStateOnEntry = helper.getStateOnEntry(newState)
+    return {
```

Comment:
If `actionResult.success` is false, but `throwOnError` is false, we then return hasError: false?

Reply:
Yes, since the developer specifically chose that this action should not throw an error we ignore all errors. I just included this config variable to try and support anything that may come up, but I doubt it will be used much since you likely only want to continue if your actions were successful.